### PR TITLE
Add source code for Chapter 18

### PR DESCRIPTION
**Pull request by Peter-VanWeert**

### CODE REVIEW

File: Exercises/NoModules/Chapter 18/Soln18_03/Customer.cpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      std::string_view surname,
+      std::string_view name,
+      std::string_view street,
+      int streetNumber,
+      std::string_view city)
+  : m_surname(surname)
+  , m_name(name)
+  , m_street(street)
+  , m_streetNumber(streetNumber)
+  , m_city(city)
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 18/Soln18_03/Customer.cpp:15]: uniform init

File: Exercises/NoModules/Chapter 18/Soln18_03/Customer.cpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  result += m_surname;
+  result += ' ';
+  result += m_name;
+  result += ", ";
+  result += m_street;
+  result += ' ';
+  result += std::to_string(m_streetNumber);
+  result += ", ";
+  result += m_city;
+  return result;
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 18/Soln18_03/Customer.cpp:30]: Actually, format would be ideal here no?
Also in other solutions.

File: Exercises/NoModules/Chapter 18/Soln18_03/DB.cpp [hunk 3]
```diff
@@ -0,0 +1,113 @@
+// A mocked database
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 18/Soln18_03/DB.cpp:1]: check the comments I had in the other chapter.
Also for all the other files in this folder.

File: Exercises/NoModules/Chapter 18/Soln18_04/Customer.cpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
+// A simple C++ customer class
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 18/Soln18_04/Customer.cpp:1]: Check with previous comments for all these files.

File: Exercises/NoModules/Chapter 18/Soln18_04/DB_RAII.h [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    }
+  }
+
+  // Prevent copying by deleting both copy members
+  DBConnectionRAII(const DBConnectionRAII& other) = delete;
+  DBConnectionRAII& operator=(const DBConnectionRAII& other) = delete;
+
+  // Allow moving by adding the appropriate members
+  DBConnectionRAII(DBConnectionRAII&& other) noexcept
+    : m_connection(other.m_connection)
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 18/Soln18_04/DB_RAII.h:44]: uniform init

File: Exercises/NoModules/Chapter 18/Soln18_05/Customer.cpp [hunk 6]
```diff
@@ -0,0 +1,31 @@
+// A simple C++ customer class
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 18/Soln18_05/Customer.cpp:1]: Check previous comments for all these files.

File: Exercises/NoModules/Chapter 18/Soln18_05/DB_RAII.h [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    }
+  }
+
+  // Prevent copying by deleting both copy members
+  DBConnectionRAII(const DBConnectionRAII& other) = delete;
+  DBConnectionRAII& operator=(const DBConnectionRAII& other) = delete;
+
+  // Allow moving by adding the appropriate members
+  DBConnectionRAII(DBConnectionRAII&& other) noexcept
+    : m_connection(std::exchange(other.m_connection, nullptr))
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 18/Soln18_05/DB_RAII.h:44]: uniform init

**Review by MarcGregoire — CHANGES_REQUESTED** [review]
Note: I mentioned the use of std::exchange() on a couple of places.
You can ignore that. I see that you introduce std::exchange() in the very last exercises in this chapter.

File: Exercises/NoModules/Chapter 18/Soln18_01/Box.h [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  Box() = default;
+  Box(double length, double width, double height) 
+    : m_length{length}, m_width{width}, m_height{height} {};
+
+  double volume() const 
+  {
+    return m_length * m_width * m_height;
+  }
+
+  friend std::ostream& operator<<(std::ostream& out, const Box& box)
```
[the remark this answers is not in the document]
Comment (Peter-VanWeert) [reply Exercises/NoModules/Chapter 18/Soln18_01/Box.h:19]: Nope. This one was a friend.

File: Exercises/NoModules/Chapter 18/Soln18_02/LinkedList.h [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+Given that this is a class template the might be used by legacy, non-moveable types,
+a perfectly sensible alternative would be to provide both overloads:
+
+  void push_front(const T& value);  // Add an object to the head (object is copied)
+  void push_back(const T& value);   // Add an object to the tail (object is copied)
+
+  void push_front(T&& value);       // Add an object to the head (object is moved)
+  void push_back(T&& value);        // Add an object to the tail (object is moved)
+
+Do not forget to the Node constructor to prevent any unnecessary copies there!
```
[the remark this answers is not in the document]
Comment (Peter-VanWeert) [reply Exercises/NoModules/Chapter 18/Soln18_02/LinkedList.h:26]: I forgot?

File: Exercises/NoModules/Chapter 18/Soln18_02/LinkedList.h [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+  LinkedList(const LinkedList& list);      // Copy constructor
+  LinkedList& operator=(const LinkedList& list); // Copy assignment operator
+
+  LinkedList(LinkedList&& list) noexcept;  // Move constructor
+  LinkedList& operator=(LinkedList&& list) noexcept; // Move assignment operator
+
+  void push_front(const T& value);         // Add an object to the head
+  void push_back(const T& value);          // Add an object to the tail
+  void push_front(T&& value);              // Move an object to the head
+  void push_back(T&& value);               // Move an object to the tail
```
[the remark this answers is not in the document]
Comment (Peter-VanWeert) [reply Exercises/NoModules/Chapter 18/Soln18_02/LinkedList.h:59]: Intro was outdated. Updated now.

**Pull request merged**